core/server/master/src/main/java/alluxio/master/metastore/rocks/RocksBlockMetaStore.java

```diff
@@ -378,21 +426,47 @@ public void addLocation(long id, BlockLocation location) {
   @Override
   public void removeLocation(long blockId, long workerId) {
     byte[] key = RocksUtils.toByteArray(blockId, workerId);
-    try {
+    try (LockResource lock = checkAndAcquireReadLock()) {
       db().delete(mBlockLocationsColumn.get(), mDisableWAL, key);
     } catch (RocksDBException e) {
       throw new RuntimeException(e);
     }
   }
 
   @Override
+  /**
+   * This iterator is used in:
+   * 1. {@link BlockIntegrityChecker} to iterate all existing blocks
+   * 2. Journal dumping like checkpoint/backup sequences
+   */
   public CloseableIterator<Block> getCloseableIterator() {
-    RocksIterator iterator = db().newIterator(mBlockMetaColumn.get(), mIteratorOption);
-    return RocksUtils.createCloseableIterator(iterator,
-        (iter) -> new Block(Longs.fromByteArray(iter.key()), BlockMeta.parseFrom(iter.value())));
+    try (LockResource lock = checkAndAcquireReadLock()) {
+      RocksIterator iterator = db().newIterator(mBlockMetaColumn.get(), mIteratorOption);
+      return RocksUtils.createCloseableIterator(iterator,
+          (iter) -> new Block(Longs.fromByteArray(iter.key()), BlockMeta.parseFrom(iter.value())),
+          mClosed::get, this::checkAndAcquireReadLock);
+    }
   }
 
   private RocksDB db() {
     return mRocksStore.getDb();
   }
+
+  // TODO(jiacheng): double check what happens if max lock count error here
+  private LockResource checkAndAcquireReadLock() {
+    // Check before locking so if the RocksDB will be closed, abort early
```

Comment:
This is what I came up for absolute correctness...
```
1. check
2. readLock
3. check again
```
I think it's hard to get rid of either check because when you grab the read lock, it may be before or after a writer.